File: web_version/pmugg_files/network/facePlacement.js
ms.facePlacement = function(normal, id, settings, face) {
	this.normal = normal;
	this.face = face;
	this.id = id;
	this.settings = settings;
	this.free = true;
	this.vertexIds = [];
	// Already existing neighboring faces that are fixed to a position.
	this.fixedNeighbors = [];
	this.d = 0;
	this.fixed = false;
	this.slope = null;
	this.value = null;
};

ms.facePlacement.prototype.getFace = function() {
	return this.face;
};

// d is the position of the plane. The plane is defined as ax + by = cz = d.
ms.facePlacement.prototype.getD = function() {
	return this.d;
};

ms.facePlacement.prototype.isFree = function() {
	return this.free;
};

ms.facePlacement.prototype.getNormal = function() {
	return this.normal;
};

ms.facePlacement.prototype.setFixed = function(fixed) {
	this.fixed = fixed;
};

ms.facePlacement.prototype.getFixed = function() {
	return this.fixed;
};

ms.facePlacement.prototype.addVertexId = function(id) {
	if (!this.vertexIds.includes(id)) {
		this.vertexIds.push(id);
	}
};

ms.facePlacement.prototype.addFixedNeighbor = function(neighbor) {
	ms.union(this.fixedNeighbors, [neighbor]);
};

ms.facePlacement.prototype.coplanar = function(fPlaceB) {
	var nA = this.getNormal();
	var nB = fPlaceB.getNormal();
	return Math.abs(nA.dot(nB)) > 1 - 1e-4;
};

ms.facePlacement.prototype.constrain = function(addBasis, vertexId) {
	var { settings } = this;
	settings.addToOrder(this.id, 'face', vertexId);

	if (addBasis) {
		settings.basisIds.push(this.id);
	}
	this.free = false;
	var self = this;
	this.vertexIds.forEach((id) => {
		var vPlace = settings.getVertex(id);
		vPlace.constrainFace(self.id);
	});
	this.vertexIds.forEach((id) => {
		settings.getVertex(id).propagate();
	});
};

ms.facePlacement.prototype.setD = function(d) {
	this.d = d;
	this.slope = 0;
	this.value = d;
};

ms.facePlacement.prototype.setFromVertex = function(vertexId) {
	var { settings } = this;
	var n = this.getNormal();
	var vPlace = settings.getVertex(vertexId);	
	var d = n.dot(vPlace.getPosition());
	if (this.fixed && Math.abs(d - this.d) > 1e-4) {
	 	return false;
	}
	this.setD(d);
	return true;
};

ms.facePlacement.prototype.makeFixed = function(faceA) {
	var { settings } = this;
	var self = this;
	this.vertexIds.forEach((id) => {
		settings.getVertex(id).addFixedNeighbor(faceA);
	});
	this.constrain(true, -1);
	faceA.faceA.getGroup().connectHole(this.face.getGroup());
};

ms.facePlacement.prototype.getRange = function(vertexId) {
	var { settings } = this;
	var {lower, upper} = settings;

	var n = this.getNormal();
	var m = 1;
	var b = 0;
	if (vertexId !== -1) {
		var vPlace = settings.getVertex(vertexId);
		m = n.dot(vPlace.slope);
		b = n.dot(vPlace.value);
	}

	if (this.fixed) {
		if (m == 0) {
			// The vertex is already fixed. This face has no effect on it.
			return new ms.range(-Infinity, Infinity);
		} else {
			return new ms.range(this.value, this.value);
		}
	}
	this.slope = m;
	this.value = b;

	var lowD = 0;
	var highD = 0;
	for (var i = 0; i < 3; i++) {
		var ni = n.getValue(i);
		if (ni > 0) {
			lowD  += ni * lower[i];
			highD += ni * upper[i];
		} else {
			lowD  += ni * upper[i];
			highD += ni * lower[i];
		}
	}
	var range = new ms.range(lowD, highD);
	this.fixedNeighbors.forEach((neighbor) => {
		var rangeI = neighbor.dirBounds(n);
		range = range.intersect(rangeI);
	});
	return ms.range.transformCreate(m, b, range);
};

ms.facePlacement.prototype.getChangeMB = function() {
	return {m: this.slope, b: this.value};
};

ms.facePlacement.prototype.print = function() {
	this.face.print();
};

File: cpp_version/pmugg/network/face_placement.cpp
#include "pch.h"
#include "face_placement.h"
#include "../graph_drawing/face.h"
#include "../fragment/net_transistor_settings.h"
#include "vertex_placement.h"
#include "../util/util.h"
#include "../util/range.h"
#include <iostream>

namespace ms {

FacePlacement::FacePlacement(const Vec3& normal, int id, NetTransistorSettings* settings, Face* face)
    : normal(normal)
    , face(face)
    , id(id)
    , settings(settings) {}

Face* FacePlacement::getFace() const {
    return face;
}

double FacePlacement::getD() const {
    return d;
}

bool FacePlacement::isFree() const {
    return free;
}

const Vec3& FacePlacement::getNormal() const {
    return normal;
}

void FacePlacement::setFixed(bool fixed) {
    this->fixed = fixed;
}

bool FacePlacement::getFixed() const {
    return fixed;
}

void FacePlacement::addVertexId(int id) {
    if (std::find(vertexIds.begin(), vertexIds.end(), id) == vertexIds.end()) {
        vertexIds.push_back(id);
    }
}

void FacePlacement::addFixedNeighbor(Face* neighbor) {
    Util::union_(fixedNeighbors, {neighbor});
}

bool FacePlacement::coplanar(FacePlacement* fPlaceB) const {
    const Vec3& nA = getNormal();
    const Vec3& nB = fPlaceB->getNormal();
    return std::abs(nA.dot(nB)) > 1.0 - 1e-4;
}

void FacePlacement::constrain(bool addBasis, int vertexId) {
    settings->addToOrder(id, "face", vertexId);

    if (addBasis) {
        settings->basisIds.push_back(id);
    }
    free = false;

    for (int id : vertexIds) {
        auto* vPlace = settings->getVertex(id);
        vPlace->constrainFace(this->id);
    }
    for (int id : vertexIds) {
        settings->getVertex(id)->propagate();
    }
}

void FacePlacement::setD(double d) {
    this->d = d;
    this->slope = 0;
    this->value = d;
}

bool FacePlacement::setFromVertex(int vertexId) {
    auto* vPlace = settings->getVertex(vertexId);
    Vec3 pos = vPlace->getPosition();
    double d = normal.dot(pos);
    
    if (fixed && std::abs(d - this->d) > 1e-4) {
        return false;
    }
    
    setD(d);
    return true;
}

void FacePlacement::makeFixed(const FixedFace& fixedFace) {
    for (int id : vertexIds) {
        settings->getVertex(id)->addFixedNeighbor(fixedFace);
    }
    constrain(true, -1);
    fixedFace.faceA->getGroup()->connectHole(face->getGroup());
}

Range FacePlacement::getRange(int vertexId) {
    const auto& lower = settings->lower;
    const auto& upper = settings->upper;

    double m = 1.0;
    double b = 0.0;

    if (vertexId != -1) {
        auto* vPlace = settings->getVertex(vertexId);
        m = normal.dot(vPlace->slope);
        b = normal.dot(vPlace->value);
    }

    if (fixed) {
        if (m == 0) {
            return Range(-std::numeric_limits<double>::infinity(),
                        std::numeric_limits<double>::infinity());
        }
        return Range((double)value, (double)value);
    }

    slope = m;
    value = b;

    double lowD = 0;
    double highD = 0;
    for (int i = 0; i < 3; i++) {
        double ni = normal[i];
        if (ni > 0) {
            lowD += ni * lower[i];
            highD += ni * upper[i];
        } else {
            lowD += ni * upper[i];
            highD += ni * lower[i];
        }
    }

    Range range((double)lowD, (double)highD);
    for (auto* neighbor : fixedNeighbors) {
        Range rangeI = neighbor->dirBounds(normal);
        range = range.intersect(rangeI);
    }

    return Range::transformCreate((double)m, (double)b, range);
}

ChangeMB FacePlacement::getChangeMB() const {
    double m = slope;
    double b = value;
    return {m, b};
}

} // namespace ms 